### ISSUE TITLE
`Union` constraint to add label to row doesn't work

### ISSUE THREAD

**Issue by chexxor** [purescript/purescript#3242]
I was playing with different ways of composing type-synonymed rows together and I hit a compilation error that I can't understand.

Is this a bug in the type-checker? Perhaps the record field accessor (the `.page` code here) doesn't have access to the `Union` constraint?

``` purescript
module Main where

import Prelude

type PaginationRow = ( page :: Int )

f :: forall r. Union PaginationRow () r => Record r -> String
f r = "Page = " <> show r.page
-- Could not match type  r1  with type  ( page :: String | t0 )

g :: forall r. Union ( page :: Int ) () r => Record r -> Int
g r = r.page
--- Could not match type  r1  with type  ( page :: String | t0 )
```

**Comment by LiamGoodacre** [comment]
Was the error for `g` supposed to say `page :: Int`?

This is currently expected behaviour.  As what @natefaubion said in slack; depending on a constraint won't introduce equality proofs/coercions in the function body.

There could be a design decision to make here about functional dependencies.  Unlike in Haskell, our functional dependencies do not generate a set of global axioms.  These axioms are rules by which we could refine/substitute determined types without necessarily resolving to a particular instance - because all instances should agree on their functional dependencies.  What I'm getting at is that, in your example, `r` is fully determined, so when type checking, the compiler could know what it should be.

A problem with doing it this way is that type checking some declaration now differs based on if the compiler has had a new axiom registered or not.  Which either means delaying the checking of some declaration until either the end or a sufficient axiom has been registered, or allowing an inconsistent view of the axioms.  We would likely also need to have special support for the compiler solved type classes.

Hopefully I'm not talking rubbish.

I'll mark as enhancement as there could be a discussion here.
Thanks for bringing it up.

**Comment by garyb** [comment]
Interesting, I don't know how viable this is, but it could be pretty nice as it'd also allow us to drop the coercion functions for the various `*Equals` classes then too, right?

I guess this appears less in Haskell as the `~` constraints are a feature of the type system rather than implemented with classes.

**Comment by chexxor** [comment]
Yeah, the type of `page` in the error message was wrong - I've updated now. Sorry!

Want to paste Nate's explanation from Slack to here for easy finding later.

> I think that’s probably it though, things like `Union` don’t propagate equality information that things like 
> property access can consume. The classes can compute things, but it’s up to the user to handle that equality through some explicit coercion
> Like, given that evidence you could do `((unsafeCoerce :: Record r -> Record (foo :: Int)) r).foo` 
> and know that it’s “safe”
> It would be like if `_.foo :: forall x r. RowCons "foo" Int x r => Record r -> Int`
> and the `Union` entails having `RowCons` constraints for all of its members
> kind of messy
- @natefaubion on FP Slack yesterday

That got me thinking that it would be interesting to think about how to unify the two ways of declaring the contents of a row at the type-level - `RowCons label type x r =>` vs `(label :: type | r)`. Could the latter be sugar for the former? But, as you suggest, this might need to depend on a rewrite/redesign of fundeps relation to type-checking (if I understand you correctly :) )

When you talk about functional dependencies, I presume they're relevant here because they are used in implementation of the `Union` class: `class Union (l :: # Type) (r :: # Type) (u :: # Type) | l r -> u, r u -> l, u l -> r`
